### PR TITLE
Insertion order index

### PR DESCRIPTION
**Pull request by mmeloni**

### CODE REVIEW

File: pkg/store/store.go [hunk 1]
```diff
@@ -296,30 +300,59 @@ func (t *Store) Count(prefix schema.KeyPrefix) (count *schema.ItemsCount, err er
...
-
-	stream.Send = func(list *pb.KVList) error {
-		for _, kv := range list.Kv {
-			key = kv.Key
-			value = kv.Value
-			version = kv.Version
-			found = true
+	var refkey []byte
+	// cache reference lookup
+	if key := t.tree.rcaches[0].Get(readTs - 1); key != nil {
```
Comment (leogr) [inline pkg/store/store.go:311]: `t.tree` should be read-locked here, shouldn't?

File: pkg/store/store.go [hunk 2]
```diff
@@ -296,30 +300,59 @@ func (t *Store) Count(prefix schema.KeyPrefix) (count *schema.ItemsCount, err er
 	return
 }
 
-// fixme(leogr): need to be optmized
-func (t *Store) itemAt(readTs uint64) (version uint64, key, value []byte, err error) {
-	stream := t.db.NewStreamAt(readTs)
-	stream.ChooseKey = func(item *badger.Item) bool {
-		return item.UserMeta() != bitTreeEntry && item.Version() == readTs
+func (t *Store) itemAt(readTs uint64) (index uint64, key, value []byte, err error) {
+	if readTs > t.tree.w {
```
Comment (leogr) [inline pkg/store/store.go:304]: This seems inconsistent to me. 
`t.tree.w` represents the width of the computed Merkel tree, not the last valid item index. For a short time frame, newly inserted items can be already stored in the DB although their node in the Merkel tree have not been computed yet.

Anyhow, if you used `t.tree.w` I believe `t.tree` should be read-locked at least.

File: pkg/store/store.go (hunk 1 above)
Comment (mmeloni) [reply pkg/store/store.go:310]: Tree is now locked

File: pkg/store/store.go (hunk 2 above)
Comment (mmeloni) [reply pkg/store/store.go]: I remove the guard.
Method return all items currently stored. When the iterator miss the first element a not found error is returned.

**Comment by mmeloni** [comment]
closed with #119

**Pull request closed without merging**